include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,233 @@
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
+   */
+  auto
+  CheckVersion()  //
+      -> bool
+  {
+    std::pair<uint64_t, bool> current_version_ = GetVersion();
+    if (!GetVersion().second) return false;
+    if (current_version_.first == version_) return true;
+    return false;
+  }
+
+  /**
+   * @brief Get an exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
+  }
+
+  /**
+   * @brief Downgrade an X lock to an SIX lock.
+   *
+   * NOTE: if a thread that does not have an exclusive lock calls this function, it will
+   * corrupt an internal lock status.
+   */
+  void
+  DowngradeToSIX()
+  {
+    lock_.store(kSIXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Release an exclusive lock.
+   *
+   */
+  void
+  UnlockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
+  }
+
+  /**
+   * @brief Get a shared lock with an intent-exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockSIX()
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_relaxed) & kSIXLockMask;
+      auto desired = expected | kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_release);
+        if (cas_success) return;
+
+        expected &= kSIXLockMask;
+        desired = expected | kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Upgrade an SIX lock to an X lock.
+   *
+   * NOTE: if a thread that does not have a shared lock with an intent-exclusive lock
+   * calls this function, it will corrupt an internal lock status.
+   */
+  void
+  UpgradeToX()
+  {
+    while (true) {
+      auto expected = kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, kXLock, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected = kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
```

Comment:
ここは`fetch_add`でOKです．

```suggestion
    lock_.fetch_add(kSIXLock, std::memory_order_relaxed);
```